File: src/test/java/com/rtomyj/yugiohAPI/service/CardServiceTest.java
package com.rtomyj.yugiohAPI.service;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.junit.Assert.assertEquals;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import com.rtomyj.yugiohAPI.configuration.exception.YgoException;
import com.rtomyj.yugiohAPI.dao.database.Dao;
import com.rtomyj.yugiohAPI.helper.ServiceLayerHelper;
import com.rtomyj.yugiohAPI.model.Card;

import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.http.HttpStatus;
import org.springframework.test.context.junit.jupiter.SpringExtension;
import com.rtomyj.yugiohAPI.helper.constants.TestConstants;
import com.rtomyj.yugiohAPI.model.Card;

import org.cache2k.integration.CacheLoaderException;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.test.context.junit4.SpringRunner;


@ExtendWith(SpringExtension.class)
@SpringBootTest
public class CardServiceTest
{
	@InjectMocks
	private CardService cardService;

	@Mock
	private Dao dao;

	private Card successfulCardReceived;
	private final String testCardId = "12345678";
	private final String testCardName = "E-HERO Stratos";

	@BeforeAll
	public static void before() {
		successfulCardReceived = Card
			.builder()
			.cardID(testCardId)
			.cardName(testCardName)
			.build();
	}



	@Test
	public void testFetchingCard_FromDB_Success()
		throws YgoException
	{
		when(dao.getCardInfo(eq(testCardId)))
			.thenReturn(successfulCardReceived);


		final Card card = cardService.getCardInfo(testCardId);

		assertEquals(TestConstants.WRONG_CARD_ID_MESSAGE, testCardId, card.getCardID());
		assertEquals(TestConstants.WRONG_CARD_NAME_MESSAGE, testCardName, card.getCardName());


		verify(dao, times(1)).getCardInfo(eq(testCardId));
	}



	@Test
	public void testFetchingCardFromDB_Failure() throws YgoException
	{
		when(dao.getCardInfo(eq(testCardId)))
			.thenThrow(new YgoException());


		assertThrows(YgoException.class, () -> cardService.getCardInfo(testCardId));


		verify(dao, times(1)).getCardInfo(eq(testCardId));
	}
}

File: src/test/kotlin/com/rtomyj/skc/find/CardServiceTest.kt
package com.rtomyj.skc.find

import com.nhaarman.mockito_kotlin.eq
import com.rtomyj.skc.config.DateConfig
import com.rtomyj.skc.constant.TestConstants
import com.rtomyj.skc.constant.TestObjects
import com.rtomyj.skc.dao.BanListDao
import com.rtomyj.skc.dao.Dao
import com.rtomyj.skc.dao.ProductDao
import com.rtomyj.skc.exception.ErrorType
import com.rtomyj.skc.exception.SKCException
import com.rtomyj.skc.model.Card
import com.rtomyj.skc.skcsuggestionengine.TrafficService
import com.rtomyj.skc.util.enumeration.TrafficResourceType
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.TestInstance
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.Mockito
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.mock.mockito.MockBean
import org.springframework.test.annotation.DirtiesContext
import org.springframework.test.context.ContextConfiguration
import org.springframework.test.context.junit.jupiter.SpringExtension
import reactor.core.publisher.Mono
import reactor.test.StepVerifier
import java.lang.Thread.sleep


@ExtendWith(SpringExtension::class)
@ContextConfiguration(classes = [CardService::class, DateConfig::class])
@TestInstance(TestInstance.Lifecycle.PER_METHOD)
@DirtiesContext(classMode = DirtiesContext.ClassMode.AFTER_EACH_TEST_METHOD) // Re-creates DiffService which is needed since cache will have the card info after one of the tests executes, ruining other tests
class CardServiceTest {
  @MockBean(name = "jdbc")
  private lateinit var cardDao: Dao

  @MockBean(name = "product-jdbc")
  private lateinit var productDao: ProductDao

  @MockBean(name = "ban-list-jdbc")
  private lateinit var banListDao: BanListDao

  @MockBean
  private lateinit var trafficService: TrafficService

  @Autowired
  private lateinit var cardService: CardService

  private val successfulCardReceived: Card = TestObjects.STRATOS_CARD_FULL_TEXT

  @Nested
  inner class HappyPath {
    /**
     * Happy path - flow where cardDao is used to retrieve card from DB. Dao object is mocked.
     */
    @Test
    fun `Test Fetching Card From DB, Success`() {
      // mock calls
      Mockito
          .`when`(cardDao.getCardInfo(eq(TestConstants.STRATOS_ID)))
          .thenReturn(successfulCardReceived)
      Mockito
          .`when`(
            trafficService.submitTrafficData(TrafficResourceType.CARD, TestConstants.STRATOS_ID, TestConstants.MOCK_IP)
          )
          .thenReturn(Mono.just(""))

      // call code
      StepVerifier
          .create(cardService.getCardInfo(TestConstants.STRATOS_ID, false, TestConstants.MOCK_IP))
          .assertNext { card ->
            // assertions on return value
            Assertions.assertEquals(TestConstants.STRATOS_ID, card.cardID)
            Assertions.assertEquals(TestConstants.STRATOS_NAME, card.cardName)
          }
          .verifyComplete()

      // verify mocks are called the exact number of times expected
      sleep(200)  // sleep as traffic call is async, and we will verify the call below
      Mockito
          .verify(cardDao, Mockito.times(1))
          .getCardInfo(eq(TestConstants.STRATOS_ID))
      Mockito
          .verify(trafficService)
          .submitTrafficData(TrafficResourceType.CARD, TestConstants.STRATOS_ID, TestConstants.MOCK_IP)
    }
  }

  @Nested
  inner class UnhappyPath {
    /**
     * Unhappy path - flow where cardDao is used to retrieve card from DB. Dao object is mocked. An error occurred while fetching using Dao.
     */
    @Test
    fun `Test Fetching Card From DB, Failure`() {
      // mock calls
      Mockito
          .`when`(cardDao.getCardInfo(eq(TestConstants.ID_THAT_CAUSES_FAILURE)))
          .thenThrow(
            SKCException(
              String.format("Unable to find card in DB with ID: %s", TestConstants.ID_THAT_CAUSES_FAILURE),
              ErrorType.DB001
            )
          )
      Mockito
          .`when`(
            trafficService.submitTrafficData(
              TrafficResourceType.CARD,
              TestConstants.ID_THAT_CAUSES_FAILURE,
              TestConstants.MOCK_IP
            )
          )
          .thenReturn(Mono.just(""))

      // call code and assert throws
      StepVerifier
          .create(
            cardService.getCardInfo(
              TestConstants.ID_THAT_CAUSES_FAILURE, false, TestConstants.MOCK_IP
            )
          )
          .verifyError()
//      Assertions.assertThrows(SKCException::class.java) { todo how ot catch speicifc error from stepverififier
//
//      }

      // verify mocks are called the exact number of times expected
      sleep(200)  // sleep as traffic call is async, and we will verify the call below
      Mockito
          .verify(
            cardDao, Mockito.times(1)
          )
          .getCardInfo(eq(TestConstants.ID_THAT_CAUSES_FAILURE))
      Mockito
          .verify(trafficService)
          .submitTrafficData(TrafficResourceType.CARD, TestConstants.ID_THAT_CAUSES_FAILURE, TestConstants.MOCK_IP)
    }
  }
}